core/store/migrate/migrations/0225_log_poller_filters_add_topics_logs_per_block.sql

```diff
@@ -1,16 +1,20 @@
 -- +goose Up
 
+-- This generates a unique BIGINT for the log_poller_filters table from hashing (name, evm_chain_id, address, event, topic2, topic3, topic4).
+-- An ordinary UNIQUE CONSTRAINT can't work for this because the topics can be NULL. Any row with any column being NULL automatically satisfies the unique constraint (NULL != NULL)
+-- There are simpler ways of doing this in postgres 12, 13, 14, and especially 15. But for now, we still officially support postgresql 11 and this should be just as efficient.
+CREATE OR REPLACE FUNCTION evm.f_log_filter_row_id(name TEXT, evm_chain_id NUMERIC, address BYTEA, event BYTEA, topic2 BYTEA, topic3 BYTEA, topic4 BYTEA)
```

Comment:
This seems very complex to me, can't we just create a unique index with coalesce, sth like

```sql
CREATE UNIQUE INDEX evm_log_poller_filters_unique_idx ON evm.log_poller_filters
(name, evm_chain_id, address, event, COALESCE(topic2, '\x'), COALESCE(topic3, '\x'), COALESCE(topic4, '\x'));
```

As a placeholder, we need to use values that will never be inserted into these rows, making it a bit tricky. But on the other hand sth like '\x' or '\x000000...` seems like a reasonable default for an index. @reductionista wdyt?

Reply:
We can't use '\x000000...` because 0 and NULL need to be treated differently.  NULL is a wild card that matches any value for the topic while 0 matches only the value 0, which is a fairly common value to have for a topic (for example, if the topic is a boolean).  I think you're right, that some sort of solution with COALESCE is probably better than what I have here... but it would need to involve some type conversion, or a random value that we just hope is never needed.

Now that @samsondav has gotten rid of postgresql 11 support, there is a better way available I believe so I'm going to try that first, will keep COALESCE in mind as a backup.

Reply:
Well, it wasn't quite as simple as I was expecting in postgresql 12.

postgresql 12 provides a fast efficient 64-bit hash function `hashtextextended`, so I figured we could use `hashtextextended(textin(record_out(name, evm_chain_id, address, event, topic2, topic3, topic4)))` directly as the key in the `CREATE UNIQUE INDEX` clause.  The only problem with that is that it can't infer automatically that this is an IMMUTABLE function, and gives an error saying all INDEX keys must be immutable functions.  So we still need to define this helper function, just so we can specify that it's immutable.

However, after going back over all the options I think this is still the best postgresql 12 option. (I read through several online discussions of this problem and that seems to be the consensus.)

I think your `\x` solution might be best if it were just the issue of NULL's we had. But there's an additional problem with creating indexes on this many columns. It's not recommended due to it resulting in an enormous index size and is slow to search in.  One reference I saw online says that anything over 3 columns is usually not worth it.

Reply:
> I think your \x solution might be best if it were just the issue of NULL's we had. But there's an additional problem with creating indexes on this many columns. It's not recommended due to it resulting in an enormous index size and is slow to search in. One reference I saw online says that anything over 3 columns is usually not worth it.

Would that be the case for tables like `evm.log_poller_filters` which is supposed to be rather small? For instance, CCIP on prod-tesnet running on 10-11 chains, has 1000 records there ;) Anyway, approving, because we need that backward compatibility fix. Probably, whatever we have here would be good enough for small table ;)

Reply:
LogTriggers could have more, as they plan to support an indefinite number of custom filters for thousands of users where they each define their own.  But yeah, in general the table will be small compared to something like evm.logs.  I hadn't thought about that, you're probably right that the storage difference would end up being insignificant.  We can revisit later, if we're sure we don't need to scale too much, but glad we've got this fix in now before other migrations get added on top of it.